### DISCUSSION TITLE
The button order in the unsaved changes dialogue is problematic

### DISCUSSION THREAD

**Discussion by Ifthel** [godotengine/godot-proposals#6949] [Editor]
At least for Windows, the current order of the buttons in the unsaved changes dialog (which pops up when closing a modified file) is nonstandard and can lead to accidentally deleting work because of muscle memory. The order on Windows should be Save, then Discard, then Cancel—the Cancel button is _always_ on the right in Windows (and, afaik, always on the left in macOS).

Here's Godot's dialog on Windows (Save, then Cancel, then Discard):
![image](https://github.com/godotengine/godot-proposals/assets/18341739/12bd23d4-9379-45cd-a1ab-f46d93bf1776)

Here's the expected order for Windows programs (Save, then Discard, then Cancel):
![image](https://github.com/godotengine/godot-proposals/assets/18341739/778c1989-a2c4-4ba3-b6d6-d6fcee89e68f)
![image](https://github.com/godotengine/godot-proposals/assets/18341739/887c6bcc-5247-4c13-991a-92b395346a65)

This has caused me to lose or almost lose work on several occasions.

**Comment by Zireael07** [comment]
This may be the standard on Windows, but e.g. many Linux programs follow this standard instead: https://web.archive.org/web/20130922172031/https://developer.gnome.org/hig-book/stable/windows-alert.html.en

**Reply by thehatless** [reply]
What percentage of Godot users use windows, and what percent Linux?

**Reply by Zireael07** [reply]
No clue, anyway this should probably be handled similarly to how it's done on Mac as @AThousandShips said below

**Reply by AThousandShips** [reply]
Might be best to change the current "should swap cancel and ok" for a more elaborate "dialog button order" with 3+ standards

This could be further improved by making all alternate buttons be placed the same way (as they currently are not, some are placed between, and some like this, on windows that is)

**Reply by Ifthel** [reply]
Should also note that the button order varies on Linux; KDE seems to follow Windows's order, for example.
Broadening the "should swap cancel and ok" setting would be great since it would give users full control if the default for their system is incorrect.

**Comment by AThousandShips** [comment]
This is handled on mac btw, as in the placement of cancel and ok are swapped there (but not sure where the other buttons are placed)

The setting is [here](https://docs.godotengine.org/en/stable/classes/class_displayserver.html#class-displayserver-method-get-swap-cancel-ok)

Handled here btw (this specific dialog, not the placement of cancel/ok):
https://github.com/godotengine/godot/blob/master/editor/plugins/script_editor_plugin.cpp#L4030-L4035

**Reply by Ifthel** [reply]
Users who want to change the Cancel/OK button order for the editor can do so through Editor Settings > Interface > Editor > Accept Dialog Cancel OK Buttons (this can't change the discard button's order, however).

It only swaps Cancel and Save in the unsaved changes dialog, Discard stays on the right:
![image](https://github.com/godotengine/godot-proposals/assets/18341739/85c5bc3d-a969-48e0-a541-8bd7941f24d7)

**Reply by bruvzg** [reply]
> This is handled on mac btw, as in the placement of cancel and ok are swapped there (but not sure where the other buttons are placed)

It's more or less clear for OK / Cancel. But default macOS behavior for the unsaved document case is saving without asking and keeping a snapshot of the previous version, so not sure what order buttons should be in.

**Reply by Ifthel** [reply]
Looking at older versions of Apple software ([here](https://arstechnica.com/gadgets/2012/07/os-x-10-8/5/) and [here](https://www.betalogue.com/2011/09/05/lion-saveas-compact/)), the order seems to be Discard/Cancel/Save with Discard left-aligned to set it apart. I assume that order would also apply without the left-alignment, but not sure either.
